File: code/java/data_structures_article-14.java
public class Data_structures_article_14 {
    /**
     * Summary of fancy reversal methods for singly linked lists
     */
    public Object solve(Object... args) {
        if (args == null || args.length != 1) {
            throw new IllegalArgumentException("Expected: (ListNode head)");
        }

        Solution solution = new Solution();
        return solution.reverseList((ListNode) args[0]);
    }

    private static class Solution {

        public ListNode reverseList(ListNode head) {
            if (head == null || head.next == null) {
                return head;
            }
            ListNode last = reverseList(head.next);
            head.next.next = head;
            head.next = null;
            return last;
        }
    }
}

File: code/python/data_structures_article-14.py
"""Summary of fancy reversal methods for singly linked lists"""

from __future__ import annotations
from typing import Any


def solve(*args: Any, **kwargs: Any) -> Any:
    """Run this lesson exercise with positional inputs."""
    return _Solution().reverseList(*args, **kwargs)


class _Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        if head is None or head.next is None:
            return head
        last = self.reverseList(head.next)
        """
        """
        head.next.next = head
        """
        """
        head.next = None
        """
        """
        return last
